doc/amoc_throttle.md

```diff
@@ -0,0 +1,152 @@
+Below is a sketch of user's behaviour during a test.
+It waits for messages in loop, and sends one after receiving message `{send_message, To}`.
+Rate of messages sent during a test will not exceed a previously set one bound to the name `message_rate`.
+Sending messages is scheduled in `set_up/1` function and in user loop if some arbitrary condition is met.
+This models the behaviour common across load tests, when users respond only to some messages.
+```erlang
+set_up(Users) ->
+    [User ! {send_message, To} || User <- Users, To <- Users].
+
+user_loop() ->
+    receive
+        {send_message, To} -> 
+            send_xmpp_message(To),
+            user_loop();
+        Message ->
+            process_message(Message),
+            user_loop()       
+    end.
+
+process_message(Message) ->
+    case some_condition(Message) of
+        true -> 
+            To = get_sender(Message),
+            amoc_throttle:send(message_rate, {send_message, To}); 
+        false ->
+            ok 
+    end.
+
+```
+
+For a more comprehensive example please refer to the `throttle_test` scenario, which shows possible usages of Amoc throttle.
+
+## How it works
+
+### Module overview
+
+- `amoc_throttle.erl` - provides an API for `amoc_throttle_controller`.
+- `amoc_throttle_controller.erl` - a gen_server which is responsible for reacting to requests, and managing `throttle_processes`.
+In distributed environment an instance of `throttle_controller` runs on every node, although the one running on master Amoc node is the one keeping the state for all nodes.
+- `amoc_throttle_process.erl` - gen_server module, implements the logic responsible for limiting rate.
+For every `Name` a `NoOfProcesses` are created, each responsible for keeping executions at a level of proportional part of the `Rate`.
+
+### Distributed environment
+
+#### Metrics
+In distributed environment every Amoc node with throttle started exposes metrics showing numbers of requests and executions.
+Those exposed by master node show the sum of all metrics from all nodes.
+This allows to quickly see the real rates across the whole system.
+
+#### Workflow
+When user executes `amoc_throttle:run(Name, Fun)` a request is reported to a metric that runs on user's node.
+Then a runner process is spawned on the same node.
+It's task will be to execute `Fun` asynchronously.
+A random throttle process which is assigned to the `Name` is asked for permission for asynchronous runner to execute `Fun`.
+When the request reaches the master node, where throttle processes reside, request metric on master node is updated and throttle process which got the request starts monitoring the asynchronous runner process.
+Then, depending on the load on the system and the current rate of executions, the asynchronous runner is allowed to run the `Fun` or it has to wait, because executing the function would exceed the calculated `Rate` in an `Interval`.
```

Comment:
> Then, depending on the system's load [...]

(reordering for genitive)

Also
> [...] is allowed to run the `Fun` or compelled to wait

For consistency between two connected sentences.